2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md

```diff
@@ -0,0 +1,106 @@
+During `Import-Module` a module wil be loaded into a separete Windows PowerShell process instead of current PowerShell Core if:<br />
+(A module is located in `System32` module path) and ((`CompatiblePSEditions` is `Desktop`) or (`CompatiblePSEditions` is missing))<br />
+If this condition is detected PowerShell Core:
+  1. creates a hidden `WindowsPS Compatibility` Windows PS process (unless one already exists)
+  2. creates PS Remoting connection to its IPC pipe
+  3. loads the module in remote Windows PS process
+  4. generates local proxy module/commands using IPC remoting connection and `Import-PSSession` cmdlet
+  5. when 'WindowsPS-only' module unload request is detected - unload module in remote process, close PS remoting channel, close remote process
+
+### PS Remoting Transport
+
+IPC / Named Pipe is to be used for connections to remote PowerShell process. Today a named pipe listener is created with each Windows PowerShell process. IPC transport has good performance and secured endpoints, however (unlike, for example, SSH transport) it is not currently supported by New-PSSession.
+So a new IPC-specific parameter set will be added to New-PSSession:  `New-PSSession -ProcessId <Int32> [-Name <String>] [-AppDomainName <String>]`
+Here is the pseudo-code that shows the essence of proposed implementation of a new parameter set in New-PSSession:
+```csharp
+NamedPipeConnectionInfo connectionInfo = new NamedPipeConnectionInfo(WindowsPS_ProcessId, WindowsPS_AppDomainName);
+TypeTable typeTable = TypeTable.LoadDefaultTypeFiles();
+RemoteRunspace remoteRunspace = RunspaceFactory.CreateRunspace(connectionInfo, this.Host, typeTable) as RemoteRunspace;
+remoteRunspace.Name = NamedPipeRunspaceName;
+try
+{
+    remoteRunspace.Open();
+}
+catch (RuntimeException e)
+{
+    // handle connection errors
+}
+PSSession remotePSSession = new PSSession(remoteRunspace);
+this.RunspaceRepository.Add(remotePSSession);
+WriteObject(remotePSSession);
+// In case of 'Windows PS Compatibility' scenario, Import-PSSession is then run on remotePSSession to generate proxy cmdlets.
+```
+
+### Lifetime of 'compatibility' Windows PowerShell process and module
+
+Overall RFC goal is to have familiar 'local module' user experience even though actual operations are working with a module in a separate 'compatibility' process. This drives following:
+
+1. One 'compatibility' Windows PowerShell process corresponds to one local PowerShell Core process. E.g. if a same user creates 2 local PowerShell Core processes and loads 'WindowsPS-only' module in each one, this will result in creating 2 Windows PS processes and loading the actuall module in each one of them.

```

Comment:
There should be runspace afinity between these processes, otherwise some runspaces from Core will communicate with one runspace in Windows PowerShell that it wrong I believe.

Reply:
Valid point; In some cases it is fine; in others - it is not. Unfortunately remoting transport currently does not allow this and implementing this is a significant amount of work. I'll add your point to the Alternate Considerations to keep track of this.